src/QirRuntime/lib/Tracer/README.md

```diff
@@ -0,0 +1,177 @@
+ conditionals into `__quantum__qis__apply_if*` calls. The tracer will add operations from _both branches_ into the
+ layers it creates to compute the upper bound estimate.
+
+Nested conditionals, conditional measurements and conditional tracked operations will _not_ be supported.
+
+__Caching__ (lower priority): It might be a huge perf win if the Resource Tracer could cache statistics for repeated
+ computations. The Tracer will have an option to cache layering results per quantum module if the boundaries of modules
+ are treated as layering barriers.
+
+## Layering ##
+
+_Definition_: ___Time___ is an integer-valued function on all quantum operations in a program (gates, measurements,
+ qubits allocation/release). For each gate there is start and end times. For each qubit, there are times when the qubit
+ is allocated and released. Start time of a gate cannot be less than allocation time of any of the qubits the gate uses.
+ If two gates or measurements use the same qubit, one of the gates must have start time greater than or equal to the end
+ time of the other.
+
+A sequentially executed quantum program can be assigned a trivial time function, when all quantum operations have
+ duration of 1 and unique start times, ordered to match the flow of the program. Layering compresses the timeline by
+ assuming that some operations might be executed simultaneously while allowing for different operations to have various
+ durations.
+
+_Definition_: Provided a valid _time_ function for the program a ___layer of duration N at time T, denoted as L(T,N),___
+ is a subset of operations in the program such that all of these operations have start time greater or equal _T_ and
+ finish time less than _T + N_. The program is ___layered___ if all gates in it are partitioned into layers, that don't
+ overlap in time. The union of all qubits that are involved in operations of a given layer, will be denoted _Qubits(T,N)_.
+
+A sequential program can be trivially layered such that each layer contains exactly one operation. Notice, that the
+ definition of layer doesn't require the gates to be executed _in parallel_. For example, all gates in a fully sequential
+ program can be also placed into a single layer L(0, infinity). Some gates might be considered to be very cheap and take
+ zero time to execute, those gates can be added to a layer even if they act on the same qubit another gate in this layer
+ is acting on and have to be executed sequentially within the layer.
+
+_Definition_: A ___global barrier___ is any operation that acts on _all_ currently allocated qubits. We'll provide means
+ for the clients to inject global barriers equivalent to an identity operator. This will allow the clients to enforce a
+ particular layering structure (because no later operation can sink below the barrier).
+
+The global barriers will be implemented as calls to `__quantum__qis__global_barrier` function. No additional support from

```

Comment:
> The global barriers will be implemented as calls to __quantum__qis__global_barrier function.

Does the signature of this function matter? If yes then you might want to add the signature.
Intuitively I feel that you meant to say:
> The global barrier will be implemented **inside of the** __quantum__qis__global_barrier **callback**. **And the _injection_ of a global barrier will be done by inserting the call (to this callback) into ...**

Reply:
The full signatures are defined in the table that describes all functions, supported by the tracer. The reader can search the document to see it. I'll be reworking the definition.